### PR TITLE
NEW: add Idle finance adapters

### PR DESCRIPTION
**Pull request by bugduino**

### CODE REVIEW

File: contracts/adapters/idle/IdleAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// Copyright (C) 2020 Igor Sobolev <sobolev@zerion.io>
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleAdapter.sol]:
```suggestion
// Copyright (C) 2020 Zerion Inc. <https://zerion.io>
```

File: contracts/adapters/idle/IdleAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+pragma solidity 0.6.4;
+pragma experimental ABIEncoderV2;
+
+import { ProtocolAdapter } from "../ProtocolAdapter.sol";
+import { ERC20 } from "../../ERC20.sol";
+
+
+/**
+ * @title Adapter for idle.finance protocol.
+ * @dev Implementation of ProtocolAdapter interface.
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleAdapter.sol]: add @author with your full name and email (like it is now in master)

File: contracts/adapters/idle/IdleTokenAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2020 Igor Sobolev <sobolev@zerion.io>
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleTokenAdapter.sol]: the same is here

File: contracts/adapters/idle/IdleTokenAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import { TokenAdapter } from "../TokenAdapter.sol";
+import { TokenMetadata, Component } from "../../Structs.sol";
+import { ERC20 } from "../../ERC20.sol";
+
+
+/**
+ * @dev IdleToken contract interface.
+ * Only the functions required for IdleAdapter contract are added.
+ * The IdleToken contracts is available here
+ * github.com/bugduino/idle-contracts/contracts
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleTokenAdapter.sol]: link is broken
provide link to the IdleToken contract

File: contracts/adapters/idle/IdleTokenAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ */
+interface IdleToken {
+    function token() external view returns (address);
+    function tokenPrice() external view returns (uint256);
+}
+
+
+/**
+ * @title Adapter for IdleTokens.
+ * @dev Implementation of TokenAdapter interface.
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleTokenAdapter.sol]: add @author tag

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 6]
```diff
@@ -401,6 +407,19 @@ module.exports = async (deployer, network, accounts) => {
...
     '0',
   ]);
+  await deployer.deploy(IdleAdapter, { from: accounts[0] })
+    .then(() => {
+      adapters[13].push(IdleAdapter.address);
+      tokens[13].push(idleAdapterTokens);
+    });
+  protocolNames.push('idle.finance');
+  metadata.push([
+    'idle.finance',
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js]: this field is protocol name, may be it's better to leave 'Idle' here?

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 7]
```diff
@@ -401,6 +407,19 @@ module.exports = async (deployer, network, accounts) => {
     'protocol-icons.s3.amazonaws.com/0x-staking.png',
     '0',
   ]);
+  await deployer.deploy(IdleAdapter, { from: accounts[0] })
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js]: i think you should insert your adapter before iearn adapters

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 8]
```diff
@@ -401,6 +407,19 @@ module.exports = async (deployer, network, accounts) => {
...
+    .then(() => {
+      adapters[13].push(IdleAdapter.address);
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js]: please, add another `,[]` in `adapters` and `tokens` arrays declarations

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 9]
```diff
@@ -468,6 +493,7 @@ module.exports = async (deployer, network, accounts) => {
           'CToken',
           'Curve pool token',
           'YToken',
+          'IdleToken',
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js]: `'IdleToken'` should be before `'YToken'`

File: test/AdapterRegistryDeployed.js [hunk 10]
```diff
@@ -43,6 +44,7 @@ contract.skip('AdapterRegistry deployed', () => {
             'PoolTogether pool',
             'Chai token',
             'YToken',
+            'IdleToken',
```
Comment (sobolev-igor) [inline test/AdapterRegistryDeployed.js]: before `'YToken'`

File: contracts/adapters/idle/IdleTokenAdapter.sol [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * github.com/bugduino/idle-contracts/contracts
+ */
+interface IdleToken {
+    function token() external view returns (address);
+    function tokenPrice() external view returns (uint256);
+}
+
+
+/**
+ * @title Adapter for IdleTokens.
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleTokenAdapter.sol]: `Adapter `-> `Token adapter`

File: contracts/adapters/idle/IdleTokenAdapter.sol [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+pragma experimental ABIEncoderV2;
+
+import { TokenAdapter } from "../TokenAdapter.sol";
+import { TokenMetadata, Component } from "../../Structs.sol";
+import { ERC20 } from "../../ERC20.sol";
+
+
+/**
+ * @dev IdleToken contract interface.
+ * Only the functions required for IdleAdapter contract are added.
```
Comment (sobolev-igor) [inline contracts/adapters/idle/IdleTokenAdapter.sol]: `IdleAdapter ` -> `IdleTokenAdapter`

File: migrations_scripts/1_use_registry_and_add_adapters.js (hunk 6 above)
Comment (bugduino) [reply migrations_scripts/1_use_registry_and_add_adapters.js]: ok

**Comment by bugduino** [comment]
Thanks for the review, I just committed the proposed changes

**Comment by bugduino** [comment]
I actually didn't get the question, what you mean with "if you should return DAI as underlying asset or cDAI, iDAI, and aDAI respectively" ?

**Comment by sobolev-igor** [comment]
I'll merge the PR for now, but may be in future it will work the following way:

**IdleTokenAdapter** contract returns 3 components: cDAI, iDAI, and aDAI with the corresponding exchange rates. After that **AdapterRegistry** contract will resolve these components into DAI with the corresponding exchange rates. This will allow getting the real underlying tokens for IdleToken (whether it  uses Compound or Aave at the moment) and will still be able to get the total DAI amount.

**Comment by sobolev-igor** [comment]
Anyway, thanks a lot for the cooperation!

**Comment by bugduino** [comment]
Thanks for closing. But regarding the price the fastest way to have eg the number of DAI that corresponds to X IdleDAI, is to multiply the IdleDAI quantity for IdleDAI tokenPrice (which already takes into account all the exchange rates of all cDAI, iDAI, aDAI)

**Comment by bugduino** [comment]
Btw let me know when this will actually be used in production so we can make announce it!
Thanks

**Pull request merged**